File: bench/ts/ssh-bench/algorithmBenchmark.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Buffer } from 'buffer';
import {
	SshAlgorithms,
	EncryptionAlgorithm,
	HmacAlgorithm,
	KeyExchangeAlgorithm,
	PublicKeyAlgorithm,
	MessageSigner,
	MessageVerifier,
} from '@microsoft/dev-tunnels-ssh';
import { Benchmark } from './benchmark';

declare type hrtime = [number, number];
const millis = ([s, ns]: hrtime) => s * 1000 + ns / 1000000;

export class EncryptionBenchmark extends Benchmark {
	private static readonly EncryptDecryptTimeMeasurement = 'Encrypt+Decrypt time (ms)';
	private static readonly ThroughputMeasurement = 'Throughput (MB/s)';

	private readonly algorithm: EncryptionAlgorithm;
	private readonly payloadSize: number;

	public constructor(algorithm: EncryptionAlgorithm, payloadSize: number) {
		super(
			`Encryption ${algorithm.name} ${payloadSize}B`,
			'algorithm-encryption',
			{ algo: algorithm.name, size: payloadSize.toString() },
		);

		this.higherIsBetter.set(EncryptionBenchmark.EncryptDecryptTimeMeasurement, false);
		this.algorithm = algorithm;
		this.payloadSize = payloadSize;
	}

	public async run(): Promise<void> {
		const algorithm = this.algorithm;
		const key = Buffer.alloc(algorithm.keyLength);
		const encIv = Buffer.alloc(algorithm.blockLength);
		const decIv = Buffer.alloc(algorithm.blockLength);
		SshAlgorithms.random.getBytes(key);
		SshAlgorithms.random.getBytes(encIv);
		encIv.copy(decIv);

		// Round payload to block length
		const blockLen = algorithm.blockLength;
		let alignedSize = Math.floor(this.payloadSize / blockLen) * blockLen;
		if (alignedSize < blockLen) alignedSize = blockLen;

		const plaintext = Buffer.alloc(alignedSize);
		SshAlgorithms.random.getBytes(plaintext);

		const encCipher = await algorithm.createCipher(true, key, encIv);
		const decCipher = await algorithm.createCipher(false, key, decIv);

		const startTime: hrtime = process.hrtime();

		const ciphertext = await encCipher.transform(plaintext);

		// For GCM, copy auth tag from encryptor to decryptor
		if ('sign' in encCipher && 'verify' in decCipher) {
			const signer = encCipher as unknown as MessageSigner;
			const verifier = decCipher as unknown as MessageVerifier;
			const tag = await signer.sign(Buffer.alloc(0));
			await verifier.verify(Buffer.alloc(0), tag);
		}

		await decCipher.transform(ciphertext);

		const elapsed: hrtime = process.hrtime(startTime);

		const ms = millis(elapsed);
		this.addMeasurement(EncryptionBenchmark.EncryptDecryptTimeMeasurement, ms);

		// Skip throughput for small payloads — sub-millisecond operations produce
		// wildly noisy MB/s values due to timer resolution limits.
		if (alignedSize >= 4096) {
			const megabytes = alignedSize / (1024 * 1024);
			const seconds = ms / 1000;
			this.addMeasurement(
				EncryptionBenchmark.ThroughputMeasurement,
				seconds > 0 ? megabytes / seconds : 0,
			);
		}

		encCipher.dispose();
		decCipher.dispose();
	}

	public async verify(): Promise<void> {
		const algorithm = this.algorithm;
		const key = Buffer.alloc(algorithm.keyLength);
		const encIv = Buffer.alloc(algorithm.blockLength);
		const decIv = Buffer.alloc(algorithm.blockLength);
		SshAlgorithms.random.getBytes(key);
		SshAlgorithms.random.getBytes(encIv);
		encIv.copy(decIv);

		const blockLen = algorithm.blockLength;
		let alignedSize = Math.floor(this.payloadSize / blockLen) * blockLen;
		if (alignedSize < blockLen) alignedSize = blockLen;

		const plaintext = Buffer.alloc(alignedSize);
		SshAlgorithms.random.getBytes(plaintext);
		const original = Buffer.from(plaintext);

		const encCipher = await algorithm.createCipher(true, key, encIv);
		const decCipher = await algorithm.createCipher(false, key, decIv);

		const ciphertext = await encCipher.transform(plaintext);

		// For GCM, copy auth tag from encryptor to decryptor
		if ('sign' in encCipher && 'verify' in decCipher) {
			const signer = encCipher as unknown as MessageSigner;
			const verifier = decCipher as unknown as MessageVerifier;
			const tag = await signer.sign(Buffer.alloc(0));
			await verifier.verify(Buffer.alloc(0), tag);
		}

		// Verify ciphertext differs from plaintext
		if (ciphertext.equals(original)) {
			throw new Error('Ciphertext should differ from plaintext');
		}

		const decrypted = await decCipher.transform(ciphertext);

		// Verify decrypted matches original
		if (!decrypted.equals(original)) {
			throw new Error('Decrypted data does not match original plaintext');
		}

		encCipher.dispose();
		decCipher.dispose();
	}

	public async dispose(): Promise<void> {}
}

export class HmacBenchmark extends Benchmark {
	private static readonly SignVerifyTimeMeasurement = 'Sign+Verify time (ms)';

	private readonly algorithm: HmacAlgorithm;

	public constructor(algorithm: HmacAlgorithm) {
		super(
			`HMAC ${algorithm.name}`,
			'algorithm-hmac',
			{ algo: algorithm.name },
		);

		this.higherIsBetter.set(HmacBenchmark.SignVerifyTimeMeasurement, false);
		this.algorithm = algorithm;
	}

	public async run(): Promise<void> {
		const algorithm = this.algorithm;
		const key = Buffer.alloc(algorithm.keyLength);
		SshAlgorithms.random.getBytes(key);

		const signer = await algorithm.createSigner(key);
		const verifier = await algorithm.createVerifier(key);

		const data = Buffer.alloc(256);
		SshAlgorithms.random.getBytes(data);

		const startTime: hrtime = process.hrtime();

		const signature = await signer.sign(data);
		await verifier.verify(data, signature);

		const elapsed: hrtime = process.hrtime(startTime);

		this.addMeasurement(HmacBenchmark.SignVerifyTimeMeasurement, millis(elapsed));

		signer.dispose();
		verifier.dispose();
	}

	public async verify(): Promise<void> {
		const algorithm = this.algorithm;
		const key = Buffer.alloc(algorithm.keyLength);
		SshAlgorithms.random.getBytes(key);

		const signer = await algorithm.createSigner(key);
		const verifier = await algorithm.createVerifier(key);

		const data = Buffer.alloc(256);
		SshAlgorithms.random.getBytes(data);

		// Sign and verify should succeed
		const signature = await signer.sign(data);
		const valid = await verifier.verify(data, signature);
		if (!valid) {
			throw new Error('HMAC verification failed for valid data');
		}

		// Tampered data should fail verification
		const tampered = Buffer.from(data);
		tampered[0] ^= 0xff;
		const verifier2 = await algorithm.createVerifier(key);
		const invalid = await verifier2.verify(tampered, signature);
		if (invalid) {
			throw new Error('HMAC verification should have failed for tampered data');
		}

		signer.dispose();
		verifier.dispose();
		verifier2.dispose();
	}

	public async dispose(): Promise<void> {}
}

export class KeyExchangeBenchmark extends Benchmark {
	private static readonly KexTimeMeasurement = 'Key exchange time (ms)';

	private readonly algorithm: KeyExchangeAlgorithm;

	public constructor(algorithm: KeyExchangeAlgorithm) {
		super(
			`KEX ${algorithm.name}`,
			'algorithm-kex',
			{ algo: algorithm.name },
		);

		this.higherIsBetter.set(KeyExchangeBenchmark.KexTimeMeasurement, false);
		this.algorithm = algorithm;
	}

	public async run(): Promise<void> {
		const clientKex = this.algorithm.createKeyExchange();
		const serverKex = this.algorithm.createKeyExchange();

		const startTime: hrtime = process.hrtime();

		const clientPublic = await clientKex.startKeyExchange();
		const serverPublic = await serverKex.startKeyExchange();
		await clientKex.decryptKeyExchange(serverPublic);
		await serverKex.decryptKeyExchange(clientPublic);

		const elapsed: hrtime = process.hrtime(startTime);

		this.addMeasurement(KeyExchangeBenchmark.KexTimeMeasurement, millis(elapsed));

		clientKex.dispose();
		serverKex.dispose();
	}

	public async verify(): Promise<void> {
		const clientKex = this.algorithm.createKeyExchange();
		const serverKex = this.algorithm.createKeyExchange();

		const clientPublic = await clientKex.startKeyExchange();
		const serverPublic = await serverKex.startKeyExchange();

		const clientSecret = await clientKex.decryptKeyExchange(serverPublic);
		const serverSecret = await serverKex.decryptKeyExchange(clientPublic);

		// Both sides should derive the same shared secret
		if (!clientSecret.equals(serverSecret)) {
			throw new Error('Key exchange shared secrets do not match');
		}

		// Shared secret should not be empty
		if (clientSecret.length === 0) {
			throw new Error('Key exchange shared secret is empty');
		}

		clientKex.dispose();
		serverKex.dispose();
	}

	public async dispose(): Promise<void> {}
}

export class KeygenBenchmark extends Benchmark {
	private static readonly KeygenTimeMeasurement = 'Keygen time (ms)';

	private readonly algorithm: PublicKeyAlgorithm;
	private readonly keySizeInBits: number;

	public constructor(algorithm: PublicKeyAlgorithm, keySizeInBits: number) {
		super(
			`Keygen ${algorithm.keyAlgorithmName} ${keySizeInBits}`,
			'algorithm-keygen',
			{ algo: algorithm.keyAlgorithmName, size: keySizeInBits.toString() },
		);

		this.higherIsBetter.set(KeygenBenchmark.KeygenTimeMeasurement, false);
		this.algorithm = algorithm;
		this.keySizeInBits = keySizeInBits;
	}

	public async run(): Promise<void> {
		const startTime: hrtime = process.hrtime();

		const keyPair = await this.algorithm.generateKeyPair(this.keySizeInBits);

		const elapsed: hrtime = process.hrtime(startTime);

		this.addMeasurement(KeygenBenchmark.KeygenTimeMeasurement, millis(elapsed));

		keyPair.dispose();
	}

	public async verify(): Promise<void> {
		const keyPair = await this.algorithm.generateKeyPair(this.keySizeInBits);

		// Verify the generated key has the expected bit size.
		await verifyKeySize(keyPair, this.keySizeInBits, this.algorithm.keyAlgorithmName);

		// Sign test data and verify the signature
		const signer = this.algorithm.createSigner(keyPair);
		const verifier = this.algorithm.createVerifier(keyPair);

		const data = Buffer.alloc(64);
		SshAlgorithms.random.getBytes(data);

		const signature = await signer.sign(data);
		const valid = await verifier.verify(data, signature);
		if (!valid) {
			throw new Error('Keygen: signature verification failed for generated key');
		}

		signer.dispose();
		verifier.dispose();
		keyPair.dispose();
	}

	public async dispose(): Promise<void> {}
}

export class SignatureBenchmark extends Benchmark {
	private static readonly SignVerifyTimeMeasurement = 'Sign+Verify time (ms)';

	private readonly algorithm: PublicKeyAlgorithm;
	private readonly keySizeInBits: number;

	public constructor(algorithm: PublicKeyAlgorithm, keySizeInBits: number) {
		super(
			`Signature ${algorithm.name} ${keySizeInBits}`,
			'algorithm-signature',
			{ algo: algorithm.name, size: keySizeInBits.toString() },
		);

		this.higherIsBetter.set(SignatureBenchmark.SignVerifyTimeMeasurement, false);
		this.algorithm = algorithm;
		this.keySizeInBits = keySizeInBits;
	}

	public async run(): Promise<void> {
		const keyPair = await this.algorithm.generateKeyPair(this.keySizeInBits);
		const signer = this.algorithm.createSigner(keyPair);
		const verifier = this.algorithm.createVerifier(keyPair);

		const data = Buffer.alloc(256);
		SshAlgorithms.random.getBytes(data);

		const startTime: hrtime = process.hrtime();

		const signature = await signer.sign(data);
		await verifier.verify(data, signature);

		const elapsed: hrtime = process.hrtime(startTime);

		this.addMeasurement(SignatureBenchmark.SignVerifyTimeMeasurement, millis(elapsed));

		signer.dispose();
		verifier.dispose();
		keyPair.dispose();
	}

	public async verify(): Promise<void> {
		const keyPair = await this.algorithm.generateKeyPair(this.keySizeInBits);

		// Verify the generated key has the expected bit size.
		await verifyKeySize(keyPair, this.keySizeInBits, this.algorithm.keyAlgorithmName);

		const signer = this.algorithm.createSigner(keyPair);
		const verifier = this.algorithm.createVerifier(keyPair);

		const data = Buffer.alloc(256);
		SshAlgorithms.random.getBytes(data);

		// Sign and verify should succeed
		const signature = await signer.sign(data);
		const valid = await verifier.verify(data, signature);
		if (!valid) {
			throw new Error('Signature verification failed for valid data');
		}

		// Verify with wrong data should fail
		const wrongData = Buffer.alloc(256);
		SshAlgorithms.random.getBytes(wrongData);
		const verifier2 = this.algorithm.createVerifier(keyPair);
		const invalid = await verifier2.verify(wrongData, signature);
		if (invalid) {
			throw new Error('Signature verification should have failed for wrong data');
		}

		signer.dispose();
		verifier.dispose();
		verifier2.dispose();
		keyPair.dispose();
	}

	public async dispose(): Promise<void> {}
}

/**
 * Verifies that a generated key pair has the expected bit size by checking
 * the SSH public key byte length. For RSA, the modulus is the dominant
 * component, so the total byte count is a reliable proxy for key size.
 */
async function verifyKeySize(
	keyPair: import('@microsoft/dev-tunnels-ssh').KeyPair,
	expectedSizeInBits: number,
	keyAlgorithmName: string,
): Promise<void> {
	const pubKeyBytes = await keyPair.getPublicKeyBytes();
	if (!pubKeyBytes) {
		throw new Error('Failed to get public key bytes for key size verification');
	}

	if (keyAlgorithmName === 'ssh-rsa') {
		// RSA public key SSH wire format: [string "ssh-rsa"] [mpint e] [mpint n]
		// Expected total: ~11 (algo) + ~7 (exponent) + 5 (mpint header) + keySizeInBits/8.
		const expectedModulusBytes = expectedSizeInBits / 8;
		const minExpectedLength = expectedModulusBytes + 10;
		const maxExpectedLength = expectedModulusBytes + 25;

		if (pubKeyBytes.length < minExpectedLength || pubKeyBytes.length > maxExpectedLength) {
			throw new Error(
				`Expected ${expectedSizeInBits}-bit RSA key (pub key ~${minExpectedLength}-${maxExpectedLength} bytes), ` +
				`but got ${pubKeyBytes.length} bytes. Key size mismatch.`,
			);
		}
	} else if (keyAlgorithmName.startsWith('ecdsa-sha2-')) {
		// ECDSA public key SSH wire format: [string algo] [string curve] [string Q]
		// Q is uncompressed: 0x04 || X || Y, so Q length = 1 + 2*ceil(bits/8).
		const coordBytes = Math.ceil(expectedSizeInBits / 8);
		const expectedQLen = 1 + 2 * coordBytes;
		const curveName = keyAlgorithmName.substring('ecdsa-sha2-'.length);
		const expectedTotal = (4 + keyAlgorithmName.length) + (4 + curveName.length) + (4 + expectedQLen);
		// Allow small tolerance for encoding variations.
		if (pubKeyBytes.length < expectedTotal - 2 || pubKeyBytes.length > expectedTotal + 2) {
			throw new Error(
				`Expected ${expectedSizeInBits}-bit ECDSA key (pub key ~${expectedTotal} bytes), ` +
				`but got ${pubKeyBytes.length} bytes. Key size mismatch.`,
			);
		}
	}
}

File: bench/cs/Ssh.Benchmark/AlgorithmBenchmark.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Algorithms;

namespace Microsoft.DevTunnels.Ssh.Benchmark;

#if NETSTANDARD2_0 || NET4
using ValueTask = System.Threading.Tasks.Task;
#endif

class EncryptionBenchmark : Benchmark
{
	private const string EncryptDecryptTimeMeasurement = "Encrypt+Decrypt time (ms)";
	private const string ThroughputMeasurement = "Throughput (MB/s)";

	private readonly EncryptionAlgorithm algorithm;
	private readonly int payloadSize;

	public EncryptionBenchmark(EncryptionAlgorithm algorithm, int payloadSize)
		: base(
			$"Encryption {algorithm.Name} {payloadSize}B",
			"algorithm-encryption",
			new Dictionary<string, string>
			{
				{ "algo", algorithm.Name },
				{ "size", payloadSize.ToString() },
			})
	{
		HigherIsBetter[EncryptDecryptTimeMeasurement] = false;
		this.algorithm = algorithm;
		this.payloadSize = payloadSize;
	}

	protected override Task RunAsync(Stopwatch stopwatch)
	{
		var key = new Buffer(algorithm.KeyLength);
		var encIv = new Buffer(algorithm.BlockLength);
		var decIv = new Buffer(algorithm.BlockLength);
		SshAlgorithms.Random.GetBytes(key);
		SshAlgorithms.Random.GetBytes(encIv);
		encIv.CopyTo(decIv);

		// Round payload to block length
		var blockLen = algorithm.BlockLength;
		var alignedSize = (payloadSize / blockLen) * blockLen;
		if (alignedSize < blockLen) alignedSize = blockLen;

		var plaintext = new Buffer(alignedSize);
		SshAlgorithms.Random.GetBytes(plaintext);
		var ciphertext = new Buffer(alignedSize);
		var decrypted = new Buffer(alignedSize);

		using var encCipher = algorithm.CreateCipher(true, key, encIv);
		using var decCipher = algorithm.CreateCipher(false, key, decIv);

		stopwatch.Restart();

		encCipher.Transform(plaintext, ciphertext);

#if SSH_ENABLE_AESGCM
		// For GCM, copy tag from encryptor to decryptor
		if (encCipher is IMessageSigner signer && decCipher is IMessageVerifier verifier)
		{
			var tag = new Buffer(signer.DigestLength);
			signer.Sign(plaintext, tag);
			verifier.Verify(ciphertext, tag);
		}
#endif

		decCipher.Transform(ciphertext, decrypted);

		stopwatch.Stop();

		AddMeasurement(EncryptDecryptTimeMeasurement, stopwatch.Elapsed.TotalMilliseconds);

		// Skip throughput for small payloads — sub-millisecond operations produce
		// wildly noisy MB/s values due to timer resolution limits.
		if (alignedSize >= 4096)
		{
			double megabytes = (double)alignedSize / (1024 * 1024);
			double seconds = stopwatch.Elapsed.TotalSeconds;
			AddMeasurement(ThroughputMeasurement, seconds > 0 ? megabytes / seconds : 0);
		}

		return Task.CompletedTask;
	}

	public override Task VerifyAsync()
	{
		var key = new Buffer(algorithm.KeyLength);
		var encIv = new Buffer(algorithm.BlockLength);
		var decIv = new Buffer(algorithm.BlockLength);
		SshAlgorithms.Random.GetBytes(key);
		SshAlgorithms.Random.GetBytes(encIv);
		encIv.CopyTo(decIv);

		var blockLen = algorithm.BlockLength;
		var alignedSize = (payloadSize / blockLen) * blockLen;
		if (alignedSize < blockLen) alignedSize = blockLen;

		var plaintext = new Buffer(alignedSize);
		SshAlgorithms.Random.GetBytes(plaintext);
		var original = plaintext.ToArray();
		var ciphertext = new Buffer(alignedSize);
		var decrypted = new Buffer(alignedSize);

		using var encCipher = algorithm.CreateCipher(true, key, encIv);
		using var decCipher = algorithm.CreateCipher(false, key, decIv);

		encCipher.Transform(plaintext, ciphertext);

		// Check ciphertext differs from plaintext
		if (ciphertext.SequenceEqual(plaintext))
			throw new Exception("Ciphertext equals plaintext — encryption did nothing");

#if SSH_ENABLE_AESGCM
		if (encCipher is IMessageSigner signer && decCipher is IMessageVerifier verifier)
		{
			var tag = new Buffer(signer.DigestLength);
			signer.Sign(plaintext, tag);
			verifier.Verify(ciphertext, tag);
		}
#endif

		decCipher.Transform(ciphertext, decrypted);

		// Check decrypted matches original
		if (!decrypted.SequenceEqual((Buffer)original))
			throw new Exception("Decrypted data does not match original");

		return Task.CompletedTask;
	}

	public override ValueTask DisposeAsync()
	{
#if NETSTANDARD2_0 || NET4
		return Task.CompletedTask;
#else
		return ValueTask.CompletedTask;
#endif
	}
}

class HmacBenchmark : Benchmark
{
	private const string SignVerifyTimeMeasurement = "Sign+Verify time (ms)";

	private readonly HmacAlgorithm algorithm;

	public HmacBenchmark(HmacAlgorithm algorithm)
		: base(
			$"HMAC {algorithm.Name}",
			"algorithm-hmac",
			new Dictionary<string, string>
			{
				{ "algo", algorithm.Name },
			})
	{
		HigherIsBetter[SignVerifyTimeMeasurement] = false;
		this.algorithm = algorithm;
	}

	protected override Task RunAsync(Stopwatch stopwatch)
	{
		var key = new Buffer(algorithm.KeyLength);
		SshAlgorithms.Random.GetBytes(key);

		using var signer = algorithm.CreateSigner(key);
		using var verifier = algorithm.CreateVerifier(key);

		var data = new Buffer(256);
		SshAlgorithms.Random.GetBytes(data);
		var signature = new Buffer(signer.DigestLength);

		stopwatch.Restart();

		signer.Sign(data, signature);
		verifier.Verify(data, signature);

		stopwatch.Stop();

		AddMeasurement(SignVerifyTimeMeasurement, stopwatch.Elapsed.TotalMilliseconds);

		return Task.CompletedTask;
	}

	public override Task VerifyAsync()
	{
		var key = new Buffer(algorithm.KeyLength);
		SshAlgorithms.Random.GetBytes(key);

		using var signer = algorithm.CreateSigner(key);
		using var verifier = algorithm.CreateVerifier(key);

		var data = new Buffer(256);
		SshAlgorithms.Random.GetBytes(data);
		var signature = new Buffer(signer.DigestLength);

		signer.Sign(data, signature);
		if (!verifier.Verify(data, signature))
			throw new Exception("Valid signature failed verification");

		// Tamper with data
		var tampered = data.ToArray();
		tampered[0] ^= 0xFF;
		if (verifier.Verify((Buffer)tampered, signature))
			throw new Exception("Tampered data passed verification");

		return Task.CompletedTask;
	}

	public override ValueTask DisposeAsync()
	{
#if NETSTANDARD2_0 || NET4
		return Task.CompletedTask;
#else
		return ValueTask.CompletedTask;
#endif
	}
}

class KeyExchangeBenchmark : Benchmark
{
	private const string KexTimeMeasurement = "Key exchange time (ms)";

	private readonly KeyExchangeAlgorithm algorithm;

	public KeyExchangeBenchmark(KeyExchangeAlgorithm algorithm)
		: base(
			$"KEX {algorithm.Name}",
			"algorithm-kex",
			new Dictionary<string, string>
			{
				{ "algo", algorithm.Name },
			})
	{
		HigherIsBetter[KexTimeMeasurement] = false;
		this.algorithm = algorithm;
	}

	protected override Task RunAsync(Stopwatch stopwatch)
	{
		using var clientKex = algorithm.CreateKeyExchange();
		using var serverKex = algorithm.CreateKeyExchange();

		stopwatch.Restart();

		var clientPublic = clientKex.StartKeyExchange();
		var serverPublic = serverKex.StartKeyExchange();
		var clientSecret = clientKex.DecryptKeyExchange(serverPublic);
		var serverSecret = serverKex.DecryptKeyExchange(clientPublic);

		stopwatch.Stop();

		AddMeasurement(KexTimeMeasurement, stopwatch.Elapsed.TotalMilliseconds);

		return Task.CompletedTask;
	}

	public override Task VerifyAsync()
	{
		using var clientKex = algorithm.CreateKeyExchange();
		using var serverKex = algorithm.CreateKeyExchange();

		var clientPublic = clientKex.StartKeyExchange();
		var serverPublic = serverKex.StartKeyExchange();
		var clientSecret = clientKex.DecryptKeyExchange(serverPublic);
		var serverSecret = serverKex.DecryptKeyExchange(clientPublic);

		if (clientSecret == null || clientSecret.Count == 0)
			throw new Exception("Client shared secret is empty");
		if (!clientSecret.SequenceEqual(serverSecret))
			throw new Exception("Shared secrets differ");

		return Task.CompletedTask;
	}

	public override ValueTask DisposeAsync()
	{
#if NETSTANDARD2_0 || NET4
		return Task.CompletedTask;
#else
		return ValueTask.CompletedTask;
#endif
	}
}

class KeygenBenchmark : Benchmark
{
	private const string KeygenTimeMeasurement = "Keygen time (ms)";

	private readonly PublicKeyAlgorithm algorithm;
	private readonly int keySizeInBits;

	public KeygenBenchmark(PublicKeyAlgorithm algorithm, int keySizeInBits)
		: base(
			$"Keygen {algorithm.KeyAlgorithmName} {keySizeInBits}",
			"algorithm-keygen",
			new Dictionary<string, string>
			{
				{ "algo", algorithm.KeyAlgorithmName },
				{ "size", keySizeInBits.ToString() },
			})
	{
		HigherIsBetter[KeygenTimeMeasurement] = false;
		this.algorithm = algorithm;
		this.keySizeInBits = keySizeInBits;
	}

	protected override Task RunAsync(Stopwatch stopwatch)
	{
		stopwatch.Restart();

		using var keyPair = algorithm.GenerateKeyPair(keySizeInBits);

		// Force key materialization. .NET RSA.Create() is lazy and defers actual
		// key generation until first use; without this the measurement reads ~0.
		keyPair.GetPublicKeyBytes(algorithm.KeyAlgorithmName);

		stopwatch.Stop();

		AddMeasurement(KeygenTimeMeasurement, stopwatch.Elapsed.TotalMilliseconds);

		return Task.CompletedTask;
	}

	public override Task VerifyAsync()
	{
		using var keyPair = algorithm.GenerateKeyPair(keySizeInBits);

		// Verify the generated key has the expected bit size.
		KeySizeVerification.VerifyKeySize(keyPair, keySizeInBits, algorithm.KeyAlgorithmName);

		using var signer = algorithm.CreateSigner(keyPair);
		using var verifier = algorithm.CreateVerifier(keyPair);

		var data = new Buffer(256);
		SshAlgorithms.Random.GetBytes(data);
		var signature = new Buffer(signer.DigestLength);

		signer.Sign(data, signature);
		if (!verifier.Verify(data, signature))
			throw new Exception("Generated key cannot verify its own signature");

		return Task.CompletedTask;
	}

	public override ValueTask DisposeAsync()
	{
#if NETSTANDARD2_0 || NET4
		return Task.CompletedTask;
#else
		return ValueTask.CompletedTask;
#endif
	}
}

class SignatureBenchmark : Benchmark
{
	private const string SignVerifyTimeMeasurement = "Sign+Verify time (ms)";

	private readonly PublicKeyAlgorithm algorithm;
	private readonly int keySizeInBits;

	public SignatureBenchmark(PublicKeyAlgorithm algorithm, int keySizeInBits)
		: base(
			$"Signature {algorithm.Name} {keySizeInBits}",
			"algorithm-signature",
			new Dictionary<string, string>
			{
				{ "algo", algorithm.Name },
				{ "size", keySizeInBits.ToString() },
			})
	{
		HigherIsBetter[SignVerifyTimeMeasurement] = false;
		this.algorithm = algorithm;
		this.keySizeInBits = keySizeInBits;
	}

	protected override Task RunAsync(Stopwatch stopwatch)
	{
		using var keyPair = algorithm.GenerateKeyPair(keySizeInBits);
		using var signer = algorithm.CreateSigner(keyPair);
		using var verifier = algorithm.CreateVerifier(keyPair);

		var data = new Buffer(256);
		SshAlgorithms.Random.GetBytes(data);
		var signature = new Buffer(signer.DigestLength);

		stopwatch.Restart();

		signer.Sign(data, signature);
		verifier.Verify(data, signature);

		stopwatch.Stop();

		AddMeasurement(SignVerifyTimeMeasurement, stopwatch.Elapsed.TotalMilliseconds);

		return Task.CompletedTask;
	}

	public override Task VerifyAsync()
	{
		using var keyPair = algorithm.GenerateKeyPair(keySizeInBits);

		// Verify the generated key has the expected bit size.
		KeySizeVerification.VerifyKeySize(keyPair, keySizeInBits, algorithm.KeyAlgorithmName);

		using var signer = algorithm.CreateSigner(keyPair);
		using var verifier = algorithm.CreateVerifier(keyPair);

		var data = new Buffer(256);
		SshAlgorithms.Random.GetBytes(data);
		var signature = new Buffer(signer.DigestLength);

		signer.Sign(data, signature);
		if (!verifier.Verify(data, signature))
			throw new Exception("Valid signature failed verification");

		var wrongData = new Buffer(256);
		SshAlgorithms.Random.GetBytes(wrongData);
		if (verifier.Verify(wrongData, signature))
			throw new Exception("Wrong data passed verification");

		return Task.CompletedTask;
	}

	public override ValueTask DisposeAsync()
	{
#if NETSTANDARD2_0 || NET4
		return Task.CompletedTask;
#else
		return ValueTask.CompletedTask;
#endif
	}
}

/// <summary>
/// Verifies that a generated key pair has the expected bit size by inspecting
/// the SSH public key bytes. For RSA, checks that the public key byte length
/// is consistent with the expected modulus size.
/// </summary>
static class KeySizeVerification
{
	public static void VerifyKeySize(IKeyPair keyPair, int expectedSizeInBits, string keyAlgorithmName)
	{
		var pubKeyBytes = keyPair.GetPublicKeyBytes(keyAlgorithmName);

		if (keyAlgorithmName == "ssh-rsa")
		{
			// RSA public key SSH wire format: [string "ssh-rsa"] [mpint e] [mpint n]
			// The modulus n determines the key size. Expected total byte count:
			//   ~11 (algo string) + ~7 (exponent) + 5 (mpint header) + keySizeInBits/8.
			int expectedModulusBytes = expectedSizeInBits / 8;
			int minExpectedLength = expectedModulusBytes + 10;
			int maxExpectedLength = expectedModulusBytes + 25;

			if (pubKeyBytes.Count < minExpectedLength || pubKeyBytes.Count > maxExpectedLength)
			{
				throw new Exception(
					$"Expected {expectedSizeInBits}-bit RSA key (pub key ~{minExpectedLength}-{maxExpectedLength} bytes), " +
					$"but got {pubKeyBytes.Count} bytes. Key size mismatch.");
			}
		}
		else if (keyAlgorithmName.StartsWith("ecdsa-sha2-"))
		{
			// ECDSA public key SSH wire format: [string algo] [string curve] [string Q]
			// Q is uncompressed: 0x04 || X || Y, so Q length = 1 + 2*ceil(bits/8).
			int coordBytes = (expectedSizeInBits + 7) / 8;
			int expectedQLen = 1 + 2 * coordBytes;
			// Total: 4+algo.Length + 4+curve.Length + 4+Q.Length
			int algoLen = keyAlgorithmName.Length;
			string curveName = keyAlgorithmName.Substring("ecdsa-sha2-".Length);
			int expectedTotal = (4 + algoLen) + (4 + curveName.Length) + (4 + expectedQLen);
			// Allow small tolerance for encoding variations.
			if (pubKeyBytes.Count < expectedTotal - 2 || pubKeyBytes.Count > expectedTotal + 2)
			{
				throw new Exception(
					$"Expected {expectedSizeInBits}-bit ECDSA key (pub key ~{expectedTotal} bytes), " +
					$"but got {pubKeyBytes.Count} bytes. Key size mismatch.");
			}
		}
	}
}
